### ISSUE TITLE
snlite scripts for geopandas and shapely

### ISSUE THREAD

**Issue by zeffii** [Marcus-Richmond/sverchok-gis-nodes#16]
maybe it makes sense to add a folder for scripts specifically for digging through these geometry layers. But for now i'll add them here..
```python
"""
>in geom s
out verts v
out edges s
"""
def offset(inlist, n):
    return [[a+n, b+n] for a, b in inlist]

try:
    import pandas as pd
    import shapely.geometry as sg
    from sverchok.data_structure import get_edge_loop

    for geo in geom[0]:
        coords = geo.__geo_interface__['coordinates']

        if isinstance(geo, sg.Polygon):
            array_2d = np.array(coords[0])
            edge_list = get_edge_loop(array_2d.shape[0])
        elif isinstance(geo, sg.MultiPolygon):
            array_2d = np.array(list(pd.core.common.flatten(coords))).reshape((-1, 2))
            idx = 0
            edge_list = []
            for coordset in coords[0]:
                N = len(coordset)
                temp_edge_list = get_edge_loop(N)
                if idx > 0:
                    temp_edge_list = offset(temp_edge_list, idx)
                idx += N
                edge_list.extend(temp_edge_list)
        else:
            continue
        edges.append(edge_list)
        verts.append(np.c_[ array_2d, np.zeros(array_2d.shape[0])])

except Exception as err:
    print(err)
```
this appears to still miss a few multipolygon edgenets.

**Comment by zeffii** [comment]
i;d like to point out that using `>in geom s` here, the `>` allows me to drop code like ` if geom and geom[0]` (like you will have seen in earlier versions of this script..)

**Comment by zeffii** [comment]
now ... i suppose generating ngons for the 2d map is "trivial", and worth starting with

**Comment by zeffii** [comment]
![image](https://user-images.githubusercontent.com/619340/165520591-00583b19-a8fe-480a-b55c-019789872be5.png)
```python
"""
>in geom s
out verts v
out edges s
out faces s
"""

try:
    import mathutils
    import pandas as pd
    import shapely.geometry as sg
    import shapely.ops as sops
    from sverchok.data_structure import get_edge_loop
    from sverchok_gis.utils.shapely_parsing import get_edge_loops_from_multipoly

    for geo in geom[0]:
        coords = geo.__geo_interface__['coordinates']
        indices = []
        if isinstance(geo, sg.Polygon):
            array_2d = np.array(coords[0])
            edge_list = get_edge_loop(array_2d.shape[0])
            indices = mathutils.geometry.tessellate_polygon(coords)
            
            
        elif isinstance(geo, sg.MultiPolygon):
            array_2d = np.array(list(pd.core.common.flatten(coords))).reshape((-1, 2))
            edge_list = get_edge_loops_from_multipoly(coords)

        else:
            continue
        edges.append(edge_list)
        verts.append(np.c_[ array_2d, np.zeros(array_2d.shape[0])])
        faces.append(indices)

except Exception as err:
    print(err)
```

**Comment by zeffii** [comment]
- for the Flat geometry this is solved (until otherwise encountered).
- making country MultiPolygons representable in 3d/spherical conditions will involve doing a more complex tessellation
    - will require even subdivision of each MultiPolygon ( possible to do the subdivsion in 2d space using a Division intensity field which is a function of the Projection stretching going on in the 2d coordinate <----- i have no idea how any of that works )

**Comment by Marcus-Richmond** [comment]
Yea, multipolygon was a beast I tried to avoid for while starting haha.

Here's another topic for a later date - correctly drawing polygons with holes in them. For instance, a (hypothetical) detailed polygon of any country will have holes in it for lake boundaries. I remember looking into this the other week but it seems like blender has to have edges connecting holes in polygons, but GIS data will just have a list of vertices for the outer edge and entire polygon, then another list for each inner hole (what QGIS calls rings). So I'm not sure if blender can draw geometry the same way QGIS can or what. Anyway, thinking aloud here, definitely a topic for another time.

**Comment by zeffii** [comment]
yeah lakes and rivers complicate things, but a polygon can't have holes. it just means a shape is filled using multiple polygons (of which the underlying edges are not drawn for the user to see) . Blender does have a Curve data type which acceps "rings" of vertices and tends do perform quite well.

**Comment by zeffii** [comment]
in 2D it is possible to convert a list of rings (or polylines) into a single "Curve" object, and you'll get something like this.
![image](https://user-images.githubusercontent.com/619340/165931408-20ee803a-1f65-4708-b1b8-5e4a26424f74.png)

rather than writing code that will be slower in python than the native Blender Curve tesselation code, it's not too painful to write a small converter for shapely.geometry.Polygons / MultiPolygons

**Comment by zeffii** [comment]
hahah, 7 years ago I came up with a similar GIS answer: https://blender.stackexchange.com/questions/39975/how-to-draw-a-postgis-polygon-with-holes-into-blender-using-python

somebody came up with another answer, which i'm not sure is valid. but worth a try.

**Comment by Marcus-Richmond** [comment]
> if you have a geodataframe which includes such ring based shapes. dump it here : ) and we'll experiment

on it, check this out: [https://github.com/Marcus-Richmond/example-datasets/blob/main/Polygon_Rings.zip](https://github.com/Marcus-Richmond/example-datasets/blob/main/Polygon_Rings.zip)

This is an example geopackage with two layers: _Canada_simple_ and _Canada_complex_. The simple layer is a single polygon layer of canada, with a couple holes representing the larger lakes. The complex layer is the same boundary with the same holes/lakes, but has additional polygons representing islands within those lakes. The complex layer is also a multipolygon layer.
